crates/rust-analyzer/src/config.rs

```diff
@@ -651,23 +654,304 @@ pub struct Config {
     workspace_roots: Vec<AbsPathBuf>,
     caps: lsp_types::ClientCapabilities,
     root_path: AbsPathBuf,
-    detached_files: Vec<AbsPathBuf>,
     snippets: Vec<Snippet>,
     visual_studio_code_version: Option<Version>,
 
     default_config: DefaultConfigData,
-    client_config: FullConfigInput,
-    user_config: GlobalLocalConfigInput,
-    #[allow(dead_code)]
+    /// Config node that obtains its initial value during the server initialization and
+    /// by receiving a `lsp_types::notification::DidChangeConfiguration`.
+    client_config: ClientConfig,
+
+    /// Path to the root configuration file. This can be seen as a generic way to define what would be `$XDG_CONFIG_HOME/rust-analyzer/rust-analyzer.toml` in Linux.
+    /// If not specified by init of a `Config` object this value defaults to :
+    ///
+    /// |Platform | Value                                 | Example                                  |
+    /// | ------- | ------------------------------------- | ---------------------------------------- |
+    /// | Linux   | `$XDG_CONFIG_HOME` or `$HOME`/.config | /home/alice/.config                      |
+    /// | macOS   | `$HOME`/Library/Application Support   | /Users/Alice/Library/Application Support |
+    /// | Windows | `{FOLDERID_RoamingAppData}`           | C:\Users\Alice\AppData\Roaming           |
+    user_config_path: VfsPath,
+
+    /// FIXME @alibektas : Change this to sth better.
+    /// Config node whose values apply to **every** Rust project.
+    user_config: Option<RatomlNode>,
+
+    /// A special file for this session whose path is set to `self.root_path.join("rust-analyzer.toml")`
+    root_ratoml_path: VfsPath,
+
+    /// This file can be used to make global changes while having only a workspace-wide scope.
+    root_ratoml: Option<RatomlNode>,
+
+    /// For every `SourceRoot` there can be at most one RATOML file.
     ratoml_files: FxHashMap<SourceRootId, RatomlNode>,
+
+    /// Clone of the value that is stored inside a `GlobalState`.
+    source_root_parent_map: Arc<FxHashMap<SourceRootId, SourceRootId>>,
+
+    /// Changes made to client and global configurations will partially not be reflected even after `.apply_change()` was called.
+    /// This field signals that the `GlobalState` should call its `update_configuration()` method.
+    should_update: bool,
 }
 
 #[derive(Clone, Debug)]
 struct RatomlNode {
-    #[allow(dead_code)]
     node: GlobalLocalConfigInput,
+    file_id: FileId,
+}
+
+#[derive(Debug, Clone, Default)]
+struct ClientConfig {
+    node: FullConfigInput,
+    detached_files: Vec<AbsPathBuf>,
+}
+
+impl Serialize for RatomlNode {
```

Comment:
Is this really needed? When would we ever serialize this (there are more serialize additions that I am unsure why they are there, also noticed we derive it for things in `_config_data!` which also doesn't seem necessary?)

Reply:
Before I made this PR I thought I would implement a lsp_ext for users to use when they think that config infra isn't doing what it is supposed to be doing. Serializing was meant for this but now I removed everything related.